### PR TITLE
Enable stricter checks for indexing expressions

### PR DESCRIPTION
**Pull request by rictic**
- [x] CHANGELOG.md not updated, internal change

This appears to have fixed several actual bugs.

### CODE REVIEW

File: src/javascript/javascript-document.ts [hunk 1]
```diff
@@ -68,7 +69,11 @@ export class JavaScriptDocument extends ParsedDocument<Node, Visitor> {
           continue;
         }
         if (callbackName in visitor) {
-          const result: VisitResult = visitor[callbackName](node, parent);
+          // TODO(rictic): is there a maintainable way to enforce the
```
Comment (justinfagnani) [inline src/javascript/javascript-document.ts]: I don't think this is necessary or possible. We're off the type rails with this kind of meta programming.

File: src/polymer/behavior-scanner.ts [hunk 2]
```diff
@@ -27,24 +27,16 @@ import {declarationPropertyHandlers, PropertyHandlers} from './declaration-prope
 import * as docs from './docs';
 import {getOrInferPrivacy, toScannedPolymerProperty} from './js-utils';
 
-interface KeyFunc<T> {
-  (value: T): any;
-}
-
-function dedupe<T>(array: T[], keyFunc: KeyFunc<T>): T[] {
-  const bucket = {};
+function dedupe<T, K>(array: T[], keyFunc: (value: T) => K): T[] {
```
Comment (justinfagnani) [inline src/polymer/behavior-scanner.ts]: seems like this should go in utils, btw

File: src/polymer/polymer-element-mixin.ts [hunk 3]
```diff
@@ -116,8 +116,13 @@ export class PolymerElementMixin extends ElementMixin implements
...
   emitPropertyMetadata(property: PolymerProperty) {
-    const polymerMetadata: any = {};
-    const polymerMetadataFields = ['notify', 'observer', 'readOnly'];
+    const polymerMetadata:
+        {notify?: boolean, observer?: string, readOnly?: boolean} = {};
+    const polymerMetadataFields = [
+      'notify' as 'notify',
+      'observer' as 'observer',
+      'readOnly' as 'readOnly'
+    ];
```
Comment (justinfagnani) [inline src/polymer/polymer-element-mixin.ts]: `]: (keyof typeof polymerMetadata)[]'`?

File: src/polymer/behavior-scanner.ts (hunk 2 above)
Comment (rictic) [inline src/polymer/behavior-scanner.ts]: true, though I hope to delete this by just turning these arrays into maps anyways

**Pull request merged**